README.md
<!-- 
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages). 

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages). 
-->

A Dart package for configuring date-repetition patterns (e.g. "bi-weekly", "the 10th of every month", "every three years on June 1st", etc.) and finding the [DateTime] objects that lie on the pattern.

## Features
  * Configure moderately complex date-repetition patterns
  * Create lists of [DateTime] objects that follow the pattern
  * Determine if any given [DateTime] object follows the configured pattern

## Getting started

Simply download the package and import it into your project to begin using.

## Usage

### Configure a [DateInterval] object (e.g. bi-weekly, starting on January 1st, 2020)
```dart
final DateInterval interval = DateInterval(
    startDate: DateTime(2020, 01, 01), // set the first day of the interval
    interval: Intervals.weekly, // set the interval type
    period: 2, // set the period (i.e. how many [interval]s between each occurrence)
);
```

### Create a list of [DateTime] objects that fall on the interval
Using the example bi-weekly object:
```dart
final Iterable<DateTime> dates = interval.getDatesThrough(DateTime(2020, 02, 01));
/*
[
    DateTime(2020, 01, 01),
    DateTime(2020, 01, 15),
    DateTime(2020, 01, 29)
]
*/
```

### Determine if any given [DateTime] lies on the interval
Using the example bi-weekly object:
```dart
final DateTime fifteenthOfJanuary = DateTime(2020, 01, 15);
final bool isOnInterval = interval.includes(fifteenthOfJanuary);
expect(isOnInterval, isTrue); // true
```

## Additional Information
Some special rules apply for the [Intervals.monthly] interval type. Since it's possible to set up an interval for the "31st of every month," but not every month has 31 days, the [DateInterval] will look for the "closest possible date."

What this means is that, if you have a rule for the "31st of every month" (see example below), it will count the 28th of February, the 31st of March, the 30th of April, etc. all as lying on the interval.

```dart
final DateInterval lastOfTheMonth = DateInterval(
  startDate: DateTime(2020, 01, 31),
  interval: Intervals.monthly,
  period: 1,
);

bool isValidDate = lastOfTheMonth.includes(DateTime(2020, 02, 29)); // leap year
// isValidDate: true

isValidDate = lastOfTheMonth.includes(DateTime(2020, 03, 31));
// isValidDate: true

isValidDate = lastOfTheMonth.includes(DateTime(2020, 04, 30));
// isValidDate: true

final DateInterval almostTheLastOfTheMonth = DateInterval(
  startDate: DateTime(2020, 01, 30),
  interval: Intervals.monthly,
  period: 1,
);

isValidDate = almostTheLastOfTheMonth.includes(DateTime(2020, 02, 29)); // leap year
// isValidDate: true

isValidDate = almostTheLastOfTheMonth.includes(DateTime(2020, 03, 31));
// isValidDate: false

isValidDate = almostTheLastOfTheMonth.includes(DateTime(2020, 04, 30));
// isValidDate: true
```

CHANGELOG.md
## 2.1.2

Upgraded to support Flutter >=3.10.2 and Dart >=3.0.2

## 2.1.1

Patches a bug that was causing "additional dates" to be skipped if their integer value was less than that of the `startDate`'s `day` value.

## 2.1.0

<ul>
    <li>
        Added support for "additional days of the month," i.e. you can now represent "the 1st and 15th of the month" (or any arbitrary set of dates) in a single `DateInterval` object
        <ul>
            <li>
                This can be controlled by setting a list of `int` values in the `additionalDaysOfTheMonth` property on the constructor
            </li>
            <li>
                These "additional days" still abide by the "closest end date" rule
            </li>
        </ul>
    </li>
</ul>

## 2.0.0

- Removed the overridden `toString` method as it wasn't really within the scope of this project and wasn't very useful
- Removed the `daysAheadOf` method in favor of Dart's `difference.inDays`
- Minor cleanup to resolve analysis problems

## 1.0.3

- Fixed a typo in the README

## 1.0.2

- Improved examples

## 1.0.1

- Updated the `toString` method to print a readable representation of the interval
- Added an examples file

## 1.0.0

_Initial Release_
This package is based off of work that was previously developed inside another project, so most of the work was done before extracting to this package, but some improvements have been made to the logic and, most of all, readability.


test/datetime_extensions_test.dart
import 'package:flutter_date_interval/src/utils/datetime_utils.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('DateTime extensions', () {
    group('The \'withZeroTime\' function', () {
      final DateTime date = DateTime(2020, 01, 01, 12, 30, 30, 500, 500);
      final DateTime withZeroTime = date.withZeroTime();

      test(
          'should return a DateTime object with the same year, month, and date',
          () {
        expect(withZeroTime.year, equals(date.year),
            reason: 'Years are different');
        expect(withZeroTime.month, equals(date.month),
            reason: 'Months are different');
        expect(withZeroTime.day, equals(date.day),
            reason: 'Days are different');
      });

      test('should return a DateTime object with zero time values', () {
        expect(withZeroTime.hour, isZero, reason: 'Hour is not zero');
        expect(withZeroTime.minute, isZero, reason: 'Minute is not zero');
        expect(withZeroTime.second, isZero, reason: 'Second in not zero');
        expect(withZeroTime.millisecond, isZero,
            reason: 'Millisecond is not zero');
        expect(withZeroTime.microsecond, isZero,
            reason: 'Microsecond is not zero');
      });
    });

    group('The \'isSameDayAs\' function', () {
      final DateTime date = DateTime(2020, 01, 30, 12, 30, 30, 500, 500);

      test(
          'should return true for a DateTime with the same year, month, and day',
          () {
        final DateTime sameDay = DateTime(date.year, date.month, date.day);
        final bool isSameDay = date.isSameDayAs(sameDay);
        expect(isSameDay, isTrue);
      });

      test(
          'should return true for the same date, even if the time is different',
          () {
        final DateTime sameDay =
            DateTime(date.year, date.month, date.day, 1, 2, 3, 4, 5);
        final bool isSameDay = date.isSameDayAs(sameDay);
        expect(isSameDay, isTrue);
      });

      test('should return false if the year is different', () {
        final DateTime differentYear =
            DateTime(date.year + 1, date.month, date.day);
        final bool isSameDay = date.isSameDayAs(differentYear);
        expect(isSameDay, isFalse);
      });

      test('should return false if the month is different', () {
        final DateTime differentMonth =
            DateTime(date.year, date.month + 1, date.day);
        final bool isSameDay = date.isSameDayAs(differentMonth);
        expect(isSameDay, isFalse);
      });

      test('should return false if the day is different', () {
        final DateTime differentDay =
            DateTime(date.year, date.month, date.day + 1);
        final bool isSameDay = date.isSameDayAs(differentDay);
        expect(isSameDay, isFalse);
      });
    });

    group('The \'isOnDailyIntervalFrom\' function', () {
      final DateTime date = DateTime(2020, 01, 01);

      void testInterval(int interval) {
        for (var i = 0; i < 1000; i++) {
          final DateTime other = date.add(Duration(days: interval * i));
          final bool isMatch = other.isOnDailyIntervalFrom(date, interval);
          expect(isMatch, isTrue);
        }
      }

      test('should match next 100 intervals', () {
        for (int i = 1; i < 100; i++) {
          testInterval(i);
        }
      });

      test('should match the same date, regardless of interval', () {
        final DateTime other = DateTime(date.year, date.month, date.day);

        for (int i = 0; i < 1000; i++) {
          final int interval = i + 1;
          final bool isMatch = other.isOnDailyIntervalFrom(date, interval);
          expect(isMatch, isTrue);
        }
      });

      test('should not match on next 1000 invalid interval dates', () {
        DateTime offsetDate = date.add(const Duration(days: 1));

        for (int i = 0; i < 1000; i++) {
          final bool isMatch = offsetDate.isOnDailyIntervalFrom(date, 2);
          expect(isMatch, isFalse);

          offsetDate = offsetDate.add(const Duration(days: 2));
        }
      });
    });

    group('The \'isOnWeeklyIntervalFrom\' function', () {
      final DateTime date = DateTime(2020, 01, 01);

      test('should match on same date, regardless of interval', () {
        final DateTime other = DateTime(date.year, date.month, date.day);

        for (int i = 0; i < 1000; i++) {
          final bool isMatch = other.isOnWeeklyIntervalFrom(date, i + 1);
          expect(isMatch, isTrue);
        }
      });

      test('should match on next 100 valid interval', () {
        for (int interval = 1; interval <= 100; interval++) {
          for (int i = 0; i < 1000; i++) {
            final DateTime other = date.add(Duration(days: (i * 7) * interval));
            final bool isMatch = other.isOnWeeklyIntervalFrom(date, interval);
            expect(isMatch, isTrue);
          }
        }
      });

      test('should not match on next 100 invalid intervals', () {
        for (int interval = 1; interval <= 100; interval++) {
          for (int i = 0; i < 1000; i++) {
            final List<DateTime> invalidDates = [
              date.add(Duration(days: ((i * 7) * interval) + 1)),
              date.add(Duration(days: ((i * 7) * interval) + 2)),
              date.add(Duration(days: ((i * 7) * interval) + 3)),
              date.add(Duration(days: ((i * 7) * interval) + 4)),
              date.add(Duration(days: ((i * 7) * interval) + 5)),
              date.add(Duration(days: ((i * 7) * interval) + 6)),
            ];

            for (final DateTime day in invalidDates) {
              expect(day.isOnWeeklyIntervalFrom(date, interval), isFalse);
            }
          }
        }
      });
    });

    group('The \'isOnMonthlyIntervalFrom\' function', () {
      final DateTime date = DateTime(2020, 01, 01);

      test('should match the same day, regardless of the interval', () {
        final DateTime other = DateTime(date.year, date.month, date.day);

        for (int i = 0; i < 100; i++) {
          final bool isMatch = other.isOnMonthlyIntervalFrom(date, i + 1);
          expect(isMatch, isTrue);
        }
      });

      test('should match next 100 intervals for all days before the 29th', () {
        for (int day = 1; day < 29; day++) {
          final DateTime startDate = DateTime(2020, 01, day);

          for (int interval = 1; interval <= 100; interval++) {
            final DateTime other = DateTime(
              startDate.year,
              startDate.month + interval,
              startDate.day,
            );
            final bool isMatch = other.isOnMonthlyIntervalFrom(
              startDate,
              interval,
            );
            expect(
              isMatch,
              isTrue,
              reason:
                  'Failed to match $other to $startDate with interval $interval',
            );
          }
        }
      });

      group('when the start date is later than the 28th', () {
        test('should match the closest possible end date', () {
          DateTime january29th = DateTime(2022, 01, 29);
          DateTime january30th = DateTime(2022, 01, 30);
          DateTime january31st = DateTime(2022, 01, 31);

          final DateTime endOfFebruary = DateTime(2022, 02, 28);
          final DateTime endOfApril = DateTime(2022, 04, 30);

          bool isMatch = endOfFebruary.isOnMonthlyIntervalFrom(january29th, 1);
          expect(isMatch, isTrue);

          isMatch = endOfFebruary.isOnMonthlyIntervalFrom(january30th, 1);
          expect(isMatch, isTrue);
          isMatch = endOfApril.isOnMonthlyIntervalFrom(january30th, 1);
          expect(isMatch, isTrue);

          isMatch = endOfFebruary.isOnMonthlyIntervalFrom(january31st, 1);
          expect(isMatch, isTrue);
          isMatch = endOfApril.isOnMonthlyIntervalFrom(january31st, 1);
          expect(isMatch, isTrue);
        });

        test('should only occur on the closest possible date', () {
          final DateTime targetDate = DateTime(2022, 01, 31);

          // test all dates before the 28th
          for (int i = 1; i < 28; i++) {
            DateTime testDate = DateTime(2022, 02, i);
            bool isMatch = testDate.isOnMonthlyIntervalFrom(targetDate, 1);
            expect(isMatch, isFalse);
          }

          // test the closest possible date
          expect(DateTime(2022, 02, 28).isOnMonthlyIntervalFrom(targetDate, 1),
              isTrue);
        });
      });

      test('should not match next 100 invalid interval dates', () {
        List<DateTime> _makeInvalidDates(int month, int day) {
          final List<DateTime> invalidDates = [];

          for (int i = 1; i < 28; i++) {
            if (i == day) {
              continue;
            }

            invalidDates.add(DateTime(2020, month, i));
          }

          return invalidDates;
        }

        for (int day = 1; day < 28; day++) {
          final DateTime startDate = DateTime(2020, 01, day);

          for (int i = 1; i <= 100; i++) {
            final List<DateTime> invalidDates = _makeInvalidDates(
              startDate.month + i,
              day,
            );
            for (final DateTime date in invalidDates) {
              final bool isMatch = date.isOnMonthlyIntervalFrom(startDate, 1);
              expect(isMatch, isFalse);
            }
          }
        }
      });
    });

    group('the \'isOnYearlyIntervalFrom\' function', () {
      final DateTime date = DateTime(2020, 01, 01);

      test('should match the same day, regardless of the interval', () {
        final DateTime other = DateTime(date.year, date.month, date.day);

        for (int i = 0; i < 1000; i++) {
          final bool isMatch = other.isOnYearlyIntervalFrom(date, i);
          expect(isMatch, isTrue);
        }
      });

      test('should match the next 100 intervals', () {
        for (int i = 1; i <= 100; i++) {
          for (int interval = 1; interval <= 1000; interval++) {
            final DateTime other =
                DateTime(date.year + (interval * i), date.month, date.day);
            final bool isMatch = other.isOnYearlyIntervalFrom(date, interval);
            expect(isMatch, isTrue);
          }
        }
      });

      test('should not match the next 100 invalid intervals', () {
        for (int i = 0; i < 1000; i++) {
          final DateTime other =
              DateTime(date.year + 1 + (i * 2), date.month, date.day);
          final bool isMatch = other.isOnYearlyIntervalFrom(date, 2);
          expect(isMatch, isFalse);
        }
      });

      test('should not match if the day of the month is different', () {
        for (int i = 0; i < 100; i++) {
          final List<DateTime> invalidDates = [
            for (int i = 2; i <= 31; i++)
              DateTime(date.year + i, date.month, i),
          ];
          for (final DateTime d in invalidDates) {
            final bool isMatch = d.isOnYearlyIntervalFrom(date, 1);
            expect(isMatch, isFalse);
          }
        }
      });

      test('should not match if the month is different', () {
        for (int i = 0; i < 100; i++) {
          final List<DateTime> invalidDates = [
            for (int i = 2; i <= 12; i++) DateTime(date.year + i, i, date.day),
          ];
          for (final DateTime d in invalidDates) {
            final bool isMatch = d.isOnYearlyIntervalFrom(date, 1);
            expect(isMatch, isFalse);
          }
        }
      });
    });

    group('the \'monthsApartFrom\' function', () {
      final DateTime startDate = DateTime(2020, 03, 01);

      test('should work for dates within the same year', () {
        final DateTime testDate = DateTime(2020, 09, 01);
        final int monthsApart = testDate.monthsApartFrom(startDate);
        expect(monthsApart, equals(6));
      });

      test('should work dates within one year apart', () {
        final DateTime testDate = DateTime(2021, 01, 01);
        final int monthsApart = testDate.monthsApartFrom(startDate);
        expect(monthsApart, equals(10));
      });

      test('should work for dates just over one year apart', () {
        final DateTime testDate = DateTime(2021, 07, 01);
        final int monthsApart = testDate.monthsApartFrom(startDate);
        expect(monthsApart, equals(16));
      });

      test('should work for dates that are two or more years apart', () {
        for (int i = 2; i <= 1000; i++) {
          final DateTime testDate =
              DateTime(2020 + i, startDate.month, startDate.day);
          final int monthsApart = testDate.monthsApartFrom(startDate);
          expect(monthsApart, equals(12 * i));
        }
      });
    });

    group('the \'through\' function', () {
      test('should capture dates in the correct order', () {
        final DateTime startDate = DateTime(2020, 01, 01);
        final DateTime endDate = DateTime(2020, 01, 31);
        final List<int> expectedDates = [];
        for (int i = 1; i <= 31; i++) {
          expectedDates.add(i);
        }
        final Iterable<DateTime> datesBetween = startDate.through(endDate);

        expect(datesBetween.length, equals(31));
        expect(datesBetween.map((DateTime d) => d.day),
            containsAllInOrder(expectedDates));
      });

      test(
          'should not return the start and end date when the "inclusive" flag is false',
          () {
        final DateTime startDate = DateTime(2020, 01, 01);
        final DateTime endDate = DateTime(2020, 01, 31);
        final List<int> expectedDates = [];
        for (int i = 2; i < 31; i++) {
          expectedDates.add(i);
        }
        final Iterable<DateTime> dates =
            startDate.through(endDate, inclusive: false);

        expect(dates.length, equals(29));
        expect(dates.map((DateTime d) => d.day),
            containsAllInOrder(expectedDates));
      });

      test(
          'should return one date if the start and end dates are the same and inclusive is true',
          () {
        final DateTime startDate = DateTime(2020, 01, 01);
        final DateTime endDate = DateTime(2020, 01, 01);
        final List<int> expectedDates = [1];
        final Iterable<DateTime> dates = startDate.through(endDate);

        expect(dates.length, equals(1));
        expect(dates.map((DateTime d) => d.day),
            containsAllInOrder(expectedDates));
      });

      test(
          'should return an empty list if the start date and end date are the same date and inclusive is false',
          () {
        final DateTime startDate = DateTime(2020, 01, 01);
        final DateTime endDate = DateTime(2020, 01, 01);
        final Iterable<DateTime> dates =
            startDate.through(endDate, inclusive: false);

        expect(dates, isEmpty);
      });

      test(
          'should properly order dates if the start and end dates are backward',
          () {
        final DateTime startDate = DateTime(2020, 01, 31);
        final DateTime endDate = DateTime(2020, 01, 01);
        final List<int> expectedDates = [];
        for (int i = 1; i <= 31; i++) {
          expectedDates.add(i);
        }
        final Iterable<DateTime> dates = startDate.through(endDate);

        expect(dates.length, equals(31));
        expect(dates.map((DateTime d) => d.day),
            containsAllInOrder(expectedDates));
      });
    });

    group('the \'isBeforeOrSameAs\' function', () {
      test('should return true when the current date is before the target date',
          () {
        final DateTime date = DateTime(2019, 12, 31);
        final DateTime other = DateTime(2020, 01, 01);
        final bool result = date.isBeforeOrSameAs(other);
        expect(result, isTrue);
      });

      test(
          'should return true when the current date is the same as the target date',
          () {
        final DateTime date = DateTime(2020, 01, 01);
        final DateTime other = DateTime(2020, 01, 01);
        final bool result = date.isBeforeOrSameAs(other);
        expect(result, isTrue);
      });

      test('should return false when the current date is after the target date',
          () {
        final DateTime date = DateTime(2020, 01, 02);
        final DateTime other = DateTime(2020, 01, 01);
        final bool result = date.isBeforeOrSameAs(other);
        expect(result, isFalse);
      });
    });

    group('the \'isAfterOrSameAs\' function', () {
      test(
          'should return false when the current date is before the target date',
          () {
        final DateTime date = DateTime(2020, 01, 01);
        final DateTime other = DateTime(2020, 01, 02);
        final bool result = date.isAfterOrSameAs(other);
        expect(result, isFalse);
      });

      test(
          'should return true when the current date is the same as the target date',
          () {
        final DateTime date = DateTime(2020, 01, 01);
        final DateTime other = DateTime(2020, 01, 01);
        final bool result = date.isAfterOrSameAs(other);
        expect(result, isTrue);
      });

      test('should return true when the current date is after the target date',
          () {
        final DateTime date = DateTime(2020, 01, 02);
        final DateTime other = DateTime(2020, 01, 01);
        final bool result = date.isAfterOrSameAs(other);
        expect(result, isTrue);
      });
    });
  });
}


lib/src/date_interval.dart
import 'intervals.dart';
import './utils/datetime_utils.dart';

/// Defines a repetition interval.
class DateInterval {
  /// Creates an instance of the [DateInterval] class.
  /// [period] must be greater than 0 or it will throw an [ArgumentError].
  DateInterval({
    DateTime? startDate,
    DateTime? endDate,
    this.period = 1,
    this.interval = Intervals.once,
    Iterable<DateTime>? skipDates,
    Iterable<int>? additionalDaysOfTheMonth,
  }) {
    if (period < 1) {
      throw ArgumentError.value(
          period, 'period', 'Period must be 1 or greater.');
    }

    if (skipDates != null) {
      this.skipDates.addAll(skipDates);
    }

    this.startDate = startDate == null
        ? DateTime.now().withZeroTime()
        : startDate.withZeroTime();
    this.endDate = endDate?.withZeroTime();

    for (final day in additionalDaysOfTheMonth ?? <int>[]) {
      if (day < 1 || day > 31) continue;
      if (additionalDates.contains(day) || day == startDate!.day) continue;

      additionalDates.add(day);
    }
  }

  /// The beginning of the interval. Determines the day-of-month (for monthly interval), day-of-week (for weekly interval).
  late final DateTime startDate;

  /// The last possible day for this interval. Any dates beyond this date will not be counted.
  late final DateTime? endDate;

  /// The frequency at which the interval occurs, i.e. how many [interval]s between each date.
  final int period;

  /// The interval at which this pattern occurs.
  final Intervals interval;

  /// Any specific dates within the interval to skip that might otherwise be counted.
  final List<DateTime> skipDates = [];

  /// Any additional "days of the month" that work alongside the start date.
  final List<int> additionalDates = [];

  /// Returns an [Iterable<DateTime>] of all valid dates between the
  /// configured [startDate] and the earliest of the configured
  /// [endDate] (if present) or the supplied [targetEndDate].
  Iterable<DateTime> getDatesThrough(DateTime targetEndDate) {
    final List<DateTime> dates = [];
    DateTime current = startDate;

    bool isValid(DateTime date) =>
        date.isBeforeOrSameAs(targetEndDate) && !_isAfterEndDate(date);

    while (isValid(current)) {
      if (!_isSkipDate(current) && !dates.contains(current)) {
        dates.add(current);
      }

      if (interval == Intervals.once) {
        break;
      }

      current = _getNextDate(current)!;
    }

    return dates;
  }

  /// Determines if the [targetDate] falls along the configured interval.
  bool includes(DateTime targetDate) {
    if (targetDate.isBefore(startDate) ||
        _isSkipDate(targetDate) ||
        _isAfterEndDate(targetDate)) {
      return false;
    }

    switch (interval) {
      case Intervals.once:
        return targetDate.isSameDayAs(startDate);
      case Intervals.daily:
        return targetDate.isOnDailyIntervalFrom(startDate, period);
      case Intervals.weekly:
        return targetDate.isOnWeeklyIntervalFrom(startDate, period);
      case Intervals.monthly:
        return targetDate.isOnMonthlyIntervalFrom(
          startDate,
          period,
          additionalDates,
        );
      case Intervals.yearly:
        return targetDate.isOnYearlyIntervalFrom(startDate, period);
    }
  }

  bool _isAfterEndDate(DateTime date) =>
      endDate != null && date.isAfter(endDate!);

  bool _isSkipDate(DateTime date) =>
      skipDates.any((DateTime skipDate) => skipDate.isSameDayAs(date));

  DateTime? _getNextDate(DateTime current) {
    switch (interval) {
      case Intervals.daily:
        return current.add(Duration(days: period));
      case Intervals.weekly:
        return current.add(Duration(days: (7 * period)));
      case Intervals.monthly:
        final possibleDates = [
          startDate.day,
          ...additionalDates,
        ]..sort((a, b) => a.compareTo(b));

        var nextDate = current.nextClosestDay(possibleDates);
        if (nextDate != null && nextDate != current) {
          return nextDate;
        }

        nextDate = DateTime(
          current.year,
          current.month + period,
          possibleDates.first,
        );

        if (nextDate.month == current.month + period + 1) {
          nextDate = DateTime(
            current.year,
            current.month + period + 1,
            0,
          );
        }

        return nextDate;
      case Intervals.yearly:
        if (current.month == DateTime.february && current.day == 29) {
          return DateTime(
            current.year + 4,
            DateTime.february,
            29,
          );
        }

        return DateTime(
          current.year + period,
          current.month,
          current.day,
        );
      case Intervals.once:
      default:
        return null;
    }
  }
}


lib/src/utils/datetime_utils.dart
import 'package:flutter/material.dart';

extension DateTimeExtensions on DateTime {
  String get dayOfWeek {
    switch (weekday) {
      case 1:
        return 'Monday';
      case 2:
        return 'Tuesday';
      case 3:
        return 'Wednesday';
      case 4:
        return 'Thursday';
      case 5:
        return 'Friday';
      case 6:
        return 'Saturday';
      case 7:
        return 'Sunday';
      default:
        return '';
    }
  }

  String get dayOfMonth {
    if (day >= 11 && day <= 13) {
      return '${day}th';
    }

    switch (day % 10) {
      case 1:
        return '${day}st';
      case 2:
        return '${day}nd';
      case 3:
        return '${day}rd';
      default:
        return '${day}th';
    }
  }

  DateTime withZeroTime() {
    return DateTime(year, month, day, 0, 0, 0, 0, 0);
  }

  bool isSameDayAs(DateTime other) {
    return year == other.year && month == other.month && day == other.day;
  }

  bool isOnDailyIntervalFrom(DateTime startDate, int interval) {
    final int numDaysApart = difference(startDate).inDays;
    return numDaysApart % interval == 0;
  }

  bool isOnWeeklyIntervalFrom(DateTime startDate, int interval) {
    final bool areSameWeekday = startDate.weekday == weekday;
    final int weeksApart = (difference(startDate).inDays / 7).floor();
    return areSameWeekday && weeksApart % interval == 0;
  }

  DateTime? nextClosestDay(List<int> dates) {
    final index = dates.indexWhere((date) => date > day);

    if (index == -1) {
      return null;
    }

    final date = DateTime(
      year,
      month,
      dates[index],
    );

    if (date.month == month + 1) {
      return DateTime(
        year,
        month + 1,
        0,
      );
    }

    return date;
  }

  bool isOnMonthlyIntervalFrom(
    DateTime startDate,
    int interval, [
    Iterable<int>? additionalDates,
  ]) {
    final bool isOnMonthlyInterval = monthsApartFrom(startDate) % interval == 0;

    if (!isOnMonthlyInterval) {
      return false;
    }

    final int lastDayOfThisMonth = DateTime(year, month + 1, 0).day;
    if (lastDayOfThisMonth < startDate.day) {
      return day == lastDayOfThisMonth;
    }

    if (additionalDates?.isNotEmpty == true) {
      return [startDate.day, ...additionalDates!].contains(day);
    }

    return startDate.day == day;
  }

  bool isOnYearlyIntervalFrom(DateTime startDate, int interval) {
    if (startDate.isSameDayAs(this)) {
      return true;
    }

    if (isBefore(startDate)) {
      return false;
    }

    return startDate.day == day &&
        startDate.month == month &&
        (year - startDate.year) % interval == 0;
  }

  int monthsApartFrom(DateTime other) {
    final DateTime first = other.isBefore(this) ? other : this;
    final DateTime last = other.isBefore(this) ? this : other;

    final int fromYear = first.year;
    final int toYear = last.year;
    final int fromMonth = first.month;
    final int toMonth = last.month;
    final int yearsApart = toYear - fromYear;

    if (yearsApart == 0) {
      return toMonth - fromMonth;
    }

    final int monthsRemainingInFirstYear = 12 - fromMonth;
    final int monthsInBetween = monthsRemainingInFirstYear + toMonth;

    if (yearsApart == 1) {
      return monthsInBetween;
    }

    return monthsInBetween + ((yearsApart - 1) * 12);
  }

  Iterable<DateTime> through(DateTime endDate, {bool inclusive = true}) {
    final List<DateTime> days = [];
    final DateTimeRange dateRange = _getOrderedDateRange(this, endDate);
    DateTime currentDate = DateTime(
      dateRange.start.year,
      dateRange.start.month,
      dateRange.start.day + 1,
    );

    if (inclusive) {
      days.add(dateRange.start);
    }

    if (!dateRange.start.isSameDayAs(dateRange.end)) {
      while (currentDate.isBefore(dateRange.end)) {
        days.add(currentDate);

        currentDate = DateTime(
          currentDate.year,
          currentDate.month,
          currentDate.day + 1,
        );
      }

      if (inclusive) {
        days.add(currentDate);
      }
    }

    return days;
  }

  bool isBeforeOrSameAs(DateTime other) {
    return other.withZeroTime().isAfter(withZeroTime()) ||
        other.withZeroTime().isSameDayAs(withZeroTime());
  }

  bool isAfterOrSameAs(DateTime other) {
    return other.withZeroTime().isBefore(withZeroTime()) ||
        other.withZeroTime().isSameDayAs(withZeroTime());
  }

  DateTimeRange _getOrderedDateRange(DateTime dateOne, DateTime dateTwo) {
    final DateTime startDate = dateOne.isBefore(dateTwo) ? dateOne : dateTwo;
    final DateTime endDate = dateOne.isBefore(dateTwo) ? dateTwo : dateOne;

    final DateTimeRange dateRange = DateTimeRange(
      start: startDate.withZeroTime(),
      end: endDate.withZeroTime(),
    );

    return dateRange;
  }
}


example/main.dart
// ignore_for_file: avoid_print

import 'package:flutter_date_interval/flutter_date_interval.dart';

void main() {
  /* One-time event */
  final DateInterval oneTimeEvent = DateInterval(
    startDate: DateTime(2020, 01, 01),
    interval: Intervals.once,
  );

  print(oneTimeEvent.toString());
  // Once on January 1, 2020

  print(oneTimeEvent.includes(DateTime(2020, 01, 01))); // true
  print(oneTimeEvent.includes(DateTime(2020, 01, 02))); // false

  print(oneTimeEvent.getDatesThrough(DateTime(2021, 01, 01)));
  // [
  //  <2020-01-01>,
  // ]

  /* Bi-weekly event */
  final DateInterval biWeekly = DateInterval(
    startDate: DateTime(2020, 01, 01),
    interval: Intervals.weekly,
    period: 2,
  );

  print(biWeekly.toString());
  // Every 2 weeks on Wednesday

  print(biWeekly.includes(DateTime(2020, 01, 01))); // true
  print(biWeekly.includes(DateTime(2020, 01, 02))); // false
  print(biWeekly.includes(DateTime(2020, 01, 15))); // true
  print(biWeekly.includes(DateTime(2020, 03, 25))); // true

  print(biWeekly.getDatesThrough(DateTime(2020, 03, 01)));
  // [
  //  <2020-01-01>,
  //  <2020-01-15>,
  //  <2020-01-29>,
  //  <2020-02-12>,
  //  <2020-02-26>,
  // ]

  /* Every three days */
  final DateInterval threeDays = DateInterval(
    startDate: DateTime(2020, 01, 01),
    interval: Intervals.daily,
    period: 3,
  );

  print(threeDays.toString());
  // Every 3 days

  print(threeDays.includes(DateTime(2020, 01, 01))); // true
  print(threeDays.includes(DateTime(2020, 01, 02))); // false
  print(threeDays.includes(DateTime(2020, 01, 03))); // false
  print(threeDays.includes(DateTime(2020, 01, 04))); // true

  print(threeDays.getDatesThrough(DateTime(2020, 02, 01)));
  // [
  //  <2020-01-01>,
  //  <2020-01-04>,
  //  <2020-01-07>,
  //  <2020-01-10>,
  //  <2020-01-13>,
  //  <2020-01-16>,
  //  <2020-01-19>,
  //  <2020-01-22>,
  //  <2020-01-25>,
  //  <2020-01-28>,
  //  <2020-01-31>,
  // ]

  /* Bi-annual event */
  final DateInterval biAnnual = DateInterval(
    startDate: DateTime(2020, 01, 01),
    interval: Intervals.monthly,
    period: 6,
  );

  print(biAnnual.toString());
  // Every 6 months on the 1st

  print(biAnnual.includes(DateTime(2020, 01, 01))); // true
  print(biAnnual.includes(DateTime(2020, 07, 01))); // true
  print(biAnnual.includes(DateTime(2021, 01, 01))); // true

  print(biAnnual.getDatesThrough(DateTime(2021, 01, 01)));
  // [
  //  <2020-01-01>,
  //  <2020-07-01>,
  //  <2021-01-01>,
  // ]
}
